File: implementations/python/balanced_segment_merge_sort.py
def on_the_fly_balanced_merge_sort(arr):
    """
    Sorts a list using the On-the-Fly Balanced Merge Strategy.
    This algorithm identifies naturally sorted segments and merges them on-the-fly
    using a stack to maintain balanced sizes, achieving O(log n) space complexity.

    :param arr: The list to sort (will be mutated).
    :return: The mutated sorted list.
    """
    if not arr or len(arr) <= 1:
        return arr

    segments_stack = []
    i = 0
    n = len(arr)

    while i < n:
        # Detect the next segment
        start = i
        segment = []
        
        # Add the first element
        segment.append(arr[i])
        i += 1
        
        # Detect if it's ascending or descending
        is_descending = False
        if i < n:
            is_descending = arr[start] > arr[i]
        
        # Continue the segment based on direction
        while i < n:
            if is_descending:
                # Descending segment
                if arr[i-1] <= arr[i]:
                    break
            else:
                # Ascending segment
                if arr[i-1] > arr[i]:
                    break
            segment.append(arr[i])
            i += 1
        
        # If descending, reverse it to make it ascending
        if is_descending:
            segment.reverse()

        # Merge with stack if needed
        current = segment
        while segments_stack and len(current) >= len(segments_stack[-1]):
            top = segments_stack.pop()
            current = merge_two_arrays(top, current)
        segments_stack.append(current)

    # Final merge of remaining segments
    while len(segments_stack) > 1:
        a = segments_stack.pop()
        b = segments_stack.pop()
        merged = merge_two_arrays(a, b)
        segments_stack.append(merged)

    # Copy back to original array
    if segments_stack:
        arr[:] = segments_stack[0]
    else:
        arr[:] = []
    return arr

def merge_two_arrays(left, right):
    """
    Merges two sorted arrays into a single sorted array.
    
    :param left: The first sorted array.
    :param right: The second sorted array.
    :return: The merged and sorted array.
    """
    if not left:
        return right[:]
    if not right:
        return left[:]
    
    result = []
    i = 0
    j = 0
    
    # Merge while both arrays have elements
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    
    # Add remaining elements from left array
    while i < len(left):
        result.append(left[i])
        i += 1
    
    # Add remaining elements from right array
    while j < len(right):
        result.append(right[j])
        j += 1
    
    return result

File: implementations/javascript/balanced_segment_merge_sort.js
/**
 * On-the-Fly Balanced Merge Sort - JavaScript Implementation
 * Author: Mario Raúl Carbonell Martínez
 * Date: November 2025
 * 
 * An adaptive sorting algorithm that identifies naturally sorted segments
 * and merges them on-the-fly using a stack-based balanced approach.
 * Achieves O(log n) space complexity and O(n log n) time complexity.
 *
 * @param {Array<number>} arr The array to sort (will be mutated).
 * @returns {Array<number>} The mutated sorted array.
 */
function onTheFlyBalancedMergeSort(arr) {
    if (!arr || arr.length <= 1) {
        return arr;
    }

    const n = arr.length;
    const segmentsStack = [];
    let i = 0;

    while (i < n) {
        const seg = detectSegmentIndices(arr, i, n);
        let currentStart = seg[0];
        let currentEnd = seg[1];
        i = currentEnd;

        while (segmentsStack.length > 0) {
            const top = segmentsStack[segmentsStack.length - 1];
            const topLen = top[1] - top[0];
            const currentLen = currentEnd - currentStart;
            if (currentLen < topLen) {
                break;
            }
            segmentsStack.pop();
            symmerge(arr, top[0], currentStart, currentEnd);
            currentStart = top[0];
        }

        segmentsStack.push([currentStart, currentEnd]);
    }

    while (segmentsStack.length > 1) {
        const a = segmentsStack.pop();
        const b = segmentsStack.pop();
        symmerge(arr, b[0], a[0], a[1]);
        segmentsStack.push([b[0], a[1]]);
    }

    return arr;
}


function reverseSlice(arr, start, end) {
    let i = start;
    let j = end - 1;
    while (i < j) {
        const tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
        i++;
        j--;
    }
}

function detectSegmentIndices(arr, start, n) {
    if (start >= n) {
        return [start, start];
    }

    let end = start + 1;
    if (end < n && arr[start] > arr[end]) {
        while (end < n && arr[end - 1] > arr[end]) {
            end++;
        }
        reverseSlice(arr, start, end);
        return [start, end];
    } else {
        while (end < n && arr[end - 1] <= arr[end]) {
            end++;
        }
        return [start, end];
    }
}

function rotateRange(arr, first, middle, last) {
    if (first >= middle || middle >= last) {
        return;
    }
    reverseSlice(arr, first, middle);
    reverseSlice(arr, middle, last);
    reverseSlice(arr, first, last);
}

function lowerBound(arr, first, last, value) {
    let left = first;
    let right = last;
    while (left < right) {
        const mid = (left + right) >> 1;
        if (arr[mid] < value) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return left;
}

function symmerge(arr, first, middle, last) {
    if (first >= middle || middle >= last) {
        return;
    }
    if (last - first === 1) {
        return;
    }
    if (last - first === 2) {
        if (arr[middle] < arr[first]) {
            const tmp = arr[first];
            arr[first] = arr[middle];
            arr[middle] = tmp;
        }
        return;
    }

    const mid1 = Math.floor((first + middle) / 2);
    const value = arr[mid1];
    const mid2 = lowerBound(arr, middle, last, value);
    const newMid = mid1 + (mid2 - middle);
    rotateRange(arr, mid1, middle, mid2);
    symmerge(arr, first, mid1, newMid);
    symmerge(arr, newMid + 1, mid2, last);
}

// Export the function for use in other modules (e.g., tests or benchmarks)
if (typeof module !== 'undefined' && module.exports) {
    module.exports = onTheFlyBalancedMergeSort;
}
